db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java

```diff
@@ -2876,4 +2937,60 @@ public JBitSet collectInnerTablesFromFlattenedOJ() {
         }
         return collected;
     }
+
+    public Map<Integer, Set<ValueNode>> collectExpressions() {
+        HashMap<Integer, Set<ValueNode>> result = new HashMap<>();
+
+        if (groupByList != null) {
+            groupByList.collectExpressions(result);
+        }
+
+        if (orderByList != null) {
+            orderByList.collectExpressions(result);
+        }
+
+        if (whereClause != null) {
+            whereClause.collectExpressions(result);
+        }
+        if (wherePredicates != null) {
+            wherePredicates.collectExpressions(result);
+        }
+
+        if (havingClause != null) {
+            havingClause.collectExpressions(result);
+        }
+
+        if (selectAggregates != null) {
+            for (AggregateNode selectAggregate : selectAggregates) {
+                selectAggregate.collectExpressions(result);
+            }
+        }
+
+        if (havingAggregates != null) {
+            for (AggregateNode havingAggregate : havingAggregates) {
+                havingAggregate.collectExpressions(result);
+            }
+        }
+        // no need to check whereAggregates, should be empty
+
+        if (windowDefinitionList != null) {
+            for (WindowNode wn : windowDefinitionList) {
+                wn.collectExpressions(result);
+            }
+        }
+
+        for (QueryTreeNode fromItem : fromList) {
+            if (fromItem instanceof ResultSetNode) {  // this is probably always true?
+                ((ResultSetNode) fromItem).collectExpressions(result);
+            }
+        }
+
+        for (ResultColumn rc : resultColumns) {
+            rc.getExpression().collectSingleExpression(result);
+        }
+
+        return result;
```

Comment:
It seems that for non-flattened correlated subquery, we are not collecting all the expressions. Here is an example:
```
create table t1 (a1 int, b1 int, c1 int, d1 int);
create index idx_t1 on t1 (a1+1, b1+2);
create table t2 (a2 int, b2 int, c2 int, d2 int);

explain select 1 from t1 --splice-properties index=idx_t1
where a1+1 not in (select a2 from t2 where b1+2 =b2);
```
In this case, I don't see the expression a1+1 being collected. This condition is in the PR node underneath SubqueryNode but above the SELECT node, so we may examine that also.

Reply:
Yes, I found this today when working on multi-column in-subquery, that `leftOperand` is not collected because it's never part of a `SelectNode`. It happens to all subqueries that have `leftOperand` and cannot be flattened, not necessarily correlated.

Without collecting it we may make wrong decisions about an index being a covering one, but actually have an expression that cannot be replaced. I enhanced this by collecting `SubqueryNode`, too. Then collect its `leftOperand` (if not null), and also `restrictionList` in the PRN directly under `SubqueryNode`. That PRN is generated on top of `SelectNode` to store the newly created predicate. I add a test for this case.